src/katgpucbf/xbgpu/engine.py

```diff
@@ -528,254 +920,45 @@ async def _receiver_loop(self) -> None:
-                    self.dump_interval_s,
-                )
-
-            old_time_s = new_time_s
-            old_timestamp = item.timestamp
-
-            item.buffer_device.get_async(self._download_command_queue, heap.buffer)
-            item.saturated.get_async(self._download_command_queue, heap.saturated)
-            event = self._download_command_queue.enqueue_marker()
-            await katsdpsigproc.resource.async_wait_for_events([event])
-
-            if not np.any(item.present_ants):
-                # All Antennas have missed data at some point, mark the entire dump missing
-                logger.warning("All Antennas had a break in data during this accumulation")
-                heap.buffer[...] = MISSING
-                incomplete_accum_counter.inc(1)
-            elif not item.present_ants.all():
-                affected_baselines = Correlation.get_baselines_for_missing_ants(item.present_ants, self.n_ants)
-                for affected_baseline in affected_baselines:
-                    # Multiply by four as each baseline (antenna pair) has four
-                    # associated correlation components (polarisation pairs).
-                    affected_baseline_index = affected_baseline * 4
-                    heap.buffer[:, affected_baseline_index : affected_baseline_index + 4, :] = MISSING
-
-                incomplete_accum_counter.inc(1)
-            # else: No F-Engines had a break in data for this accumulation
-
-            heap.timestamp = item.timestamp
-            if self.send_stream.tx_enabled:
-                # Convert timestamp for the *end* of the heap (not the start)
-                # to a UNIX time for the sensor update. NB: this should be done
-                # *before* send_heap, because that gives away ownership of the
-                # heap.
-                end_adc_timestamp = item.timestamp + self.timestamp_increment_per_accumulation
-                end_timestamp = self.time_converter.adc_to_unix(end_adc_timestamp)
-                clip_cnt_sensor = self.sensors["xeng-clip-cnt"]
-                clip_cnt_sensor.set_value(clip_cnt_sensor.value + int(heap.saturated), timestamp=end_timestamp)
-            self.send_stream.send_heap(heap)
-
-            item.reset()
-            await self._tx_free_item_queue.put(item)
-
-        await self.send_stream.send_stop_heap()
-        logger.debug("_sender_loop completed")
-
-    async def request_capture_start(self, ctx) -> None:
-        """Start transmission of this baseline-correlation-products stream."""
-        self.send_stream.tx_enabled = True
-
-    async def request_capture_stop(self, ctx) -> None:
-        """Stop transmission of this baseline-correlation-products stream."""
-        self.send_stream.tx_enabled = False
+        pipeline = self._request_pipeline(stream_name)
+        # TODO: Introduce `make_send_stream` helper function for each pipeline
+        #       to address this 'has no attribute' mypy error
```

Comment:
Not quite sure I follow. But rather than exposing the XSend object or its B equivalent directly to the Engine, what about have the base class provide an abstract function to enable or disable transmission?

Reply:
mypy, as always, is upset because it doesn't know about the `send_stream` attribute that we're accessing below.

Your suggestion is a good one indeed, and probably a bit cleaner than reaching straight into a `pipeline` and toggling a value.